File: 77.组合.java
import java.util.*;

/*
 * @lc app=leetcode.cn id=77 lang=java
 *
 * [77] 组合
 *
 * https://leetcode-cn.com/problems/combinations/description/
 *
 * algorithms
 * Medium (72.30%)
 * Likes:    230
 * Dislikes: 0
 * Total Accepted:    38.3K
 * Total Submissions: 52.3K
 * Testcase Example:  '4\n2'
 *
 * 给定两个整数 n 和 k，返回 1 ... n 中所有可能的 k 个数的组合。
 * 
 * 示例:
 * 
 * 输入: n = 4, k = 2
 * 输出:
 * [
 * ⁠ [2,4],
 * ⁠ [3,4],
 * ⁠ [2,3],
 * ⁠ [1,2],
 * ⁠ [1,3],
 * ⁠ [1,4],
 * ]
 * 
 */

// @lc code=start
class Solution {
    List<List<Integer>>res = new ArrayList<>();
    List<List<Integer>> combine(int n, int k) {
        if (k <= 0 || n <= 0) return res;
        List<Integer> track = new ArrayList<>();
        backtrack(n, k, 1, track);
        return res;
    }

    private void backtrack(int n, int k, int start, List<Integer> track) {
        // 到达树的底部
        if (k == track.size()) {
            res.add(new ArrayList<Integer>(track));
            return;
        }
        // 注意 i 从 start 开始递增
        for (int i = start; i <= n; i++) {
            // 做选择
            track.add(i);
            backtrack(n, k, i + 1, track);
            // 撤销选择
            track.remove(track.size() - 1);
        }
    }
}
// @lc code=end

File: 77.组合.cs
/*
 * @lc app=leetcode.cn id=77 lang=csharp
 *
 * [77] 组合
 *
 * https://leetcode-cn.com/problems/combinations/description/
 *
 * algorithms
 * Medium (74.58%)
 * Likes:    363
 * Dislikes: 0
 * Total Accepted:    88.6K
 * Total Submissions: 117.3K
 * Testcase Example:  '4\n2'
 *
 * 给定两个整数 n 和 k，返回 1 ... n 中所有可能的 k 个数的组合。
 * 
 * 示例:
 * 
 * 输入: n = 4, k = 2
 * 输出:
 * [
 * ⁠ [2,4],
 * ⁠ [3,4],
 * ⁠ [2,3],
 * ⁠ [1,2],
 * ⁠ [1,3],
 * ⁠ [1,4],
 * ]
 * 
 */

// @lc code=start
public class Solution {
    public IList<IList<int>> Combine(int n, int k) {
        var ans = new List<IList<int>>();
        Dfs(n, k, 1, new List<int>(), ans);
        return ans;
    }
    private void Dfs(int n, int k, int cur, List<int> pre, List<IList<int>> ans) {
        if(k == 0) {
            ans.Add(new List<int>(pre));
            return;
        }
        for(int i = cur; i<= n;i++) {
            if(n - i + 1< k) break;
            pre.Add(i);
            Dfs(n, k -1, i + 1, pre, ans);
            pre.Remove(i);
        }
    }
}
// @lc code=end

